std-bits/base/src/main/java/org/enso/base/enso_cloud/TransientHTTPResponseCache.java

```diff
@@ -0,0 +1,419 @@
+   * limits.
+   */
+  private static Optional<Long> maxFileSizeOverrideTestOnly = Optional.empty();
+
+  private static Optional<Long> maxTotalCacheSizeOverrideTestOnly = Optional.empty();
+
+  private static final Map<String, CacheEntry> cache = new HashMap<>();
+  private static final Map<String, ZonedDateTime> lastUsed = new HashMap<>();
+
+  static EnsoHttpResponse makeRequest(
+      URIWithSecrets unresolvedURI,
+      URI resolvedURI,
+      List<Pair<String, String>> resolvedHeaders,
+      RequestMaker requestMaker)
+      throws IOException, InterruptedException, ResponseTooLargeException {
+    var cacheKey = makeHashKey(resolvedURI, resolvedHeaders);
+    if (cache.containsKey(cacheKey)) {
+      return returnCachedResponse(cacheKey);
+    } else {
+      return makeRequestAndCache(
+          unresolvedURI, resolvedURI, resolvedHeaders, cacheKey, requestMaker);
+    }
+  }
+
+  private static EnsoHttpResponse returnCachedResponse(String cacheKey) throws IOException {
+    markCacheEntryUsed(cacheKey);
+    return buildEnsoHttpResponseFromCacheEntry(cache.get(cacheKey));
+  }
+
+  /**
+   * IOExceptions thrown by the HTTP request are propagated; IOExceptions thrown while storing the
+   * data in the cache are caught.
+   *
+   * <p>Only HTTP 200 responses are cached; all others are returned uncached.
+   */
+  private static EnsoHttpResponse makeRequestAndCache(
+      URIWithSecrets unresolvedURI,
+      URI resolvedURI,
+      List<Pair<String, String>> resolvedHeaders,
+      String cacheKey,
+      RequestMaker requestMaker)
+      throws InterruptedException, IOException, ResponseTooLargeException {
+    assert !cache.containsKey(cacheKey);
+
+    var response = requestMaker.run();
+
+    // Don't cache non-200 repsonses.
+    if (response.statusCode() != 200) {
+      return response;
+    }
+
+    // If we get a content-length header, make room for the new file before
+    // downloading. If the response is too large, throw
+    // ResponseTooLargeException.
+    var sizeMaybe = response.headers().firstValue("content-length").map(Long::parseLong);
+    if (sizeMaybe.isPresent()) {
+      long size = sizeMaybe.get();
+      if (size > getMaxFileSize()) {
+        throw new ResponseTooLargeException(
+            size, getMaxFileSize(), unresolvedURI.render().toString());
+      }
+      makeRoomFor(size);
+    }
+
+    try {
+      // Download the response data.
+      int ttl = calculateTTL(response.headers());
+      ZonedDateTime expiry = getNow().plus(Duration.ofSeconds(ttl));
+      int statusCode = response.statusCode();
+      var headers = response.headers();
+      String responseDataPath = downloadResponseData(cacheKey, unresolvedURI, response);
+      long size = new File(responseDataPath).length();
+
+      // Create a cache entry. Remove old files if we are over the cache size
+      // limit.
+      var cacheEntry =
+          new CacheEntry(resolvedURI, headers, responseDataPath, size, statusCode, expiry);
```

Comment:
This is leaking secrets.

`resolvedURI` comes from `resolveURI` whose comment says 
```
Gets the actual URI with all secrets resolved, so that it can be used to create a request. This
value should never be returned to Enso.
```

However, when rebuilding `EnsoHttpResponse`, we pass this uri to it and it gets to Enso - leaking all secrets that were added as query arguments to the URI.

Reply:
I think the method `makeRequestAndCache` should not be taking the `resolved*` arguments at all. It does not need them.

The resolved values are only needed to be used in `RequestMaker` which should be opaque on purpose to avoid leaking secrets (this is done correctly) and in `makeHashKey`. No other methods should really be getting the `resolved*` values.

The `EnsoHttpResponse` is supposed to get the `renderedUri` instead - that is by design to avoid leaking the secrets.

Reply:
Fixed.